domain/application/state/storage.go

```diff
@@ -118,6 +135,226 @@ VALUES ($storageToAdd.*)`, storageToAdd{})
 	return nil
 }
 
+type storageTemplate struct {
+	meta   charmStorage
+	params application.ApplicationStorageArg
+}
+
+// insertUnitStorage inserts the storage records need to record the intent
+// for the specified new unit and storage args. Records include:
+// - storage instance
+// - filesystem
+// - volume
+// - related attachment records
+// TODO(storage) - support attaching existing storage when adding a unit
+func (st *State) insertUnitStorage(
+	ctx context.Context, tx *sqlair.TX, modelType model.ModelType, storageParentDir string,
+	appUUID coreapplication.ID, unitUUID coreunit.UUID, netNodeUUID string,
+	args []application.ApplicationStorageArg, poolKinds map[string]storage.StorageKind,
+) error {
+
+	// Reduce the count of new storage created for each existing storage
+	// being attached.
+	// TODO(storage) - implement this when unit machine storage can be supported
+	// (includes ensureCharmStorageCountChange below)
+
+	templates := make([]storageTemplate, 0, len(args))
+	for _, arg := range args {
+		storageMeta, err := st.getApplicationCharmStorageByName(ctx, tx, appUUID, arg.Name)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf(
+				"charm for application %q has no storage called %q",
+				appUUID, arg.Name,
+			).Add(applicationerrors.StorageNameNotSupported)
+		} else if err != nil {
+			return errors.Errorf("getting charm storage metadata for storage name %q application %q: %w", arg.Name, appUUID, err)
+		}
+
+		if arg.Count == 0 {
+			continue
+		}
+		templates = append(templates, storageTemplate{
+			meta:   storageMeta,
+			params: arg,
+		})
+	}
+
+	app := applicationID{ID: appUUID}
+	selectCharmStmt, err := st.Prepare(
+		`SELECT &applicationCharmUUID.charm_uuid FROM application WHERE uuid = $applicationID.uuid`,
+		app, applicationCharmUUID{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Storage is either a storage type or a pool name.
+	// Get a mapping of pool name to pool UUID for any
+	// pools specified in the app storage args.
+	poolNames := make([]string, len(args))
+	for i, stor := range args {
+		poolNames[i] = stor.PoolNameOrType
+	}
+	poolsByName, err := st.loadStoragePoolUUIDByName(ctx, tx, poolNames)
+	if err != nil {
+		return errors.Errorf("loading storage pool UUIDs: %w", err)
+	}
+
+	for _, t := range templates {
+		if err := ensureCharmStorageCountChange(t.meta, 0, t.params.Count); err != nil {
+			return err
+		}
+		for i := uint64(0); i < t.params.Count; i++ {
```

Comment:
Suggestion `for range t.params.Count {}`

Reply:
ok